test/utils/asserts.hpp

```diff
@@ -23,13 +25,57 @@ MATCHER(Result, "empty result")
 
 MATCHER_P(Result, value, "")  // NOLINT(readability-redundant-string-init)
 {
+    using namespace fizzy;
+
+    // Require the arg to be TypedExecutionResult.
+    // This can be a static_assert, but just returning false and failing a test provides better
+    // location of the error.
+    using result_type = std::remove_cv_t<std::remove_reference_t<arg_type>>;
+    if constexpr (!std::is_same_v<result_type, test::TypedExecutionResult>)
+    {
+        if (result_listener->IsInterested())
+            *result_listener << "TypedExecutionResult expected";
+        return false;
+    }
```

Comment:
to avoid repeating `std::is_same_v<result_type, test::TypedExecutionResult>` below
```suggestion
    }
    else
    {
        if (arg.trapped || !arg.has_value)
            return false;
            
        if constexpr (std::is_same_v<value_type, float>)
        {
        ...
    }
```